src/style-spec/feature_filter/index.js

```diff
@@ -53,75 +54,75 @@ function isExpressionFilter(filter: any) {
     }
-    const dynamicFilter = filter === staticFilter ? true : filter;
 
+    // Compile the static component of the filter
+    const filterSpec = latest[`filter_${layerType}`];
     const compiledStaticFilter = createExpression(staticFilter, filterSpec);
-    const compiledDynamicFilter = createExpression(dynamicFilter, filterSpec);
-    if (compiledStaticFilter.result === 'error' || compiledDynamicFilter.result === 'error') {
-        const staticFilterErrors = compiledStaticFilter.result === 'error' ?
-            compiledStaticFilter.value.map(err => `${err.key}: ${err.message}`).join(', ') :
-            'None';
-
-        const dynamicFilterErrors = compiledDynamicFilter.result === 'error' ?
-            compiledDynamicFilter.value.map(err => `${err.key}: ${err.message}`).join(', ') :
-            'None';
-
-        const errorMsg = `static-filter errors:
-            ${staticFilterErrors}
-
-            dynamic-filter errors:
-            ${dynamicFilterErrors}
-        `;
-        throw new Error(errorMsg);
+
+    let filterFunc = null;
+    if (compiledStaticFilter.result === 'error') {
+        throw new Error(compiledStaticFilter.value.map(err => `${err.key}: ${err.message}`).join(', '));
     } else {
+        filterFunc = (globalProperties: GlobalProperties, feature: Feature, canonical?: CanonicalTileID) => compiledStaticFilter.value.evaluate(globalProperties, feature, {}, canonical);
+    }
+
+    // If the static component is not equal to the entire filter then we have a dynamic component
+    // Compile the dynamic component separately
+    let dynamicFilterFunc = null;
+    if (staticFilter !== filterExp) {
+        const compiledDynamicFilter = createExpression(filterExp, filterSpec);
+
+        if (compiledDynamicFilter.result === 'error') {
+            throw new Error(compiledDynamicFilter.value.map(err => `${err.key}: ${err.message}`).join(', '));
+        } else {
+            dynamicFilterFunc = (globalProperties: GlobalProperties, feature: Feature, canonical?: CanonicalTileID, refLocation?: MercatorCoordinate) => compiledDynamicFilter.value.evaluate(globalProperties, feature, {}, canonical, undefined, undefined, refLocation);
+        }
+    }
+
+    if (filterFunc) {
         const needGeometry = geometryNeeded(staticFilter);
+
         return {
-            filter: (globalProperties: GlobalProperties, feature: Feature, canonical?: CanonicalTileID) => compiledStaticFilter.value.evaluate(globalProperties, feature, {}, canonical),
-            dynamicFilter: (globalProperties: GlobalProperties, feature: Feature, canonical?: CanonicalTileID, refLocation: ?MercatorCoordinate) => compiledDynamicFilter.value.evaluate(globalProperties, feature, {}, canonical, null, null, refLocation),
+            filter: filterFunc,
+            dynamicFilter: dynamicFilterFunc ? dynamicFilterFunc : undefined,
             needGeometry
         };
+    } else {
+        // This branch cannot happen but need to keep flow happy :(
```

Comment:
It seems like Flow is dictating the structure of this whole section at the expense of readability!
Would Flow be ok with a` if (!filterFunc) {return;}` so that you could unwrap the main return statement?
Or could you make `filterFunc` a `const` and use a try / catch?

Reply:
Nope, flow cannot detect early returns statically.

Reply:
I did a force cast instead, looks slightly cleaner.